src/include/simeng/OS/SyscallHandler.hh

```diff
@@ -106,12 +109,126 @@ struct linux_dirent64 {
   char* d_name;       // Filename (null-terminated)
 };
 
-/** A Linux kernel syscall emulation implementation, which mimics the
-   responses to Linux system calls. */
+/** The types of changes that can be made to values within the process state. */
+enum class ChangeType { REPLACEMENT, INCREMENT, DECREMENT };
+
+/** A structure describing a set of changes to the process state. */
+struct ProcessStateChange {
+  /** Type of changes to be made */
+  ChangeType type;
+
+  /** Registers to modify */
+  std::vector<Register> modifiedRegisters;
+
+  /** Values to set modified registers to */
+  std::vector<RegisterValue> modifiedRegisterValues;
+
+  /** Memory address/width pairs to modify */
+  std::vector<MemoryAccessTarget> memoryAddresses;
+
+  /** Values to write to memory */
+  std::vector<RegisterValue> memoryAddressValues;
+};
+
+/** This result from a handled syscall. */
+struct SyscallResult {
+  /** Indicates whether the outcome of the syscall is fatal for the associated
+   * core and it should therefore halt. */
+  bool fatal = false;
+
+  /** Id of the syscall to aid exception handler processing. */
+  uint64_t syscallId = 0;
+
+  /** The unique ID of the core associated with the syscall. */
+  uint64_t coreId = 0;
+
+  /** Any changes to apply to the process state. */
+  ProcessStateChange stateChange = {};
+
+  /** Default copy constructor for SyscallResult. */
+  SyscallResult(const SyscallResult& res) = default;
+
+  /** Default move constructor for SyscallResult to enable copy elision whenever
+   * it is possible. */
+  SyscallResult(SyscallResult&& res) = default;
+
+  /** Default copy assignment operator for SyscallResult. */
+  SyscallResult& operator=(const SyscallResult& res) = default;
+
+  /** Default move assignment operator for SyscallResult to enable copy elision
+   * whenever it is possible. */
+  SyscallResult& operator=(SyscallResult&& res) = default;
+};
+
+/** A struct to hold information used as arguments to a syscall. */
+struct SyscallInfo {
+  /** The ID of the syscall. */
+  uint64_t syscallId = 0;
+
+  /** The unique ID of the core associated with the syscall. */
+  uint64_t coreId = 0;
+
+  /** The unique ID of the process associated with the syscall. */
+  uint64_t threadId = 0;
+
+  /** The register values used as parameters to the invoked syscall. */
+  std::array<RegisterValue, 6> registerArguments = {};
+
+  /** The register which will be updated with the return value of the processed
+   * syscall. */
+  Register ret = {0, 0};
```

Comment:
Deinfied copy and move, constructors and operators for Register

Reply:
Just a small note, no need to define copy and move constructors for RegisterValue, upon looking at it seems it could lead to segfault, I think it needs a revisit as their might unused memory in RegiserValue

Reply:
resolved for Register